src/Servers/HttpSys/src/MessagePump.cs

```diff
@@ -275,4 +228,91 @@ public void Dispose()
+                          MessagePump messagePump,
+                          int workerIndex)
+        {
+            _asyncAcceptContext = asyncAcceptContext;
+            _messagePump = messagePump;
+            _workerIndex = workerIndex;
+            _preferInlineScheduling = _messagePump._options.UnsafePreferInlineScheduling;
+        }
+
+        public void Execute()
+        {
+            _ = ExecuteAsync();
+        }
+
+        private async Task ExecuteAsync()
+        {
+            while (!_messagePump.Stopping && _workerIndex <= _messagePump._maxAccepts)
+            {
+                // Receive a request
+                RequestContext requestContext;
+                try
+                {
+                    requestContext = await _messagePump.Listener.AcceptAsync(_asyncAcceptContext);
+
+                    if (!_messagePump.Listener.ValidateRequest(requestContext))
+                    {
+                        // Dispose the request
+                        requestContext.ReleasePins();
+                        requestContext.Dispose();
+
+                        // If either of these is false then a response has already been sent to the client, so we can accept the next request
+                        continue;
+                    }
+                }
+                catch (Exception exception)
+                {
+                    Debug.Assert(_messagePump.Stopping);
+                    if (_messagePump.Stopping)
+                    {
+                        Log.AcceptErrorStopping(_messagePump._logger, exception);
+                    }
+                    else
+                    {
+                        Log.AcceptError(_messagePump._logger, exception);
+                    }
+                    continue;
+                }
+
+                try
+                {
+                    if (_preferInlineScheduling)
+                    {
+                        await requestContext.ExecuteAsync();
+                    }
+                    else
+                    {
+                        // Queue another accept before we execute the request
+                        ThreadPool.UnsafeQueueUserWorkItem(this, preferLocal: false);
+
+                        // Use this thread to start the execution of the request (avoid the double threadpool dispatch)
+                        await requestContext.ExecuteAsync();
+
+                        // We're done with this thread
+                        return;
+                    }
+                }
+                catch (Exception ex)
+                {
+                    // Request processing failed
+                    // Log the error message, release throttle and move on
+                    Log.RequestListenerProcessError(_messagePump._logger, ex);
+                }
+            }
+            Interlocked.Decrement(ref _messagePump._acceptorCounts);
+
+            _asyncAcceptContext.Dispose();
```

Comment:
These aren't called when returning on line 303. Should they be in a finally?

Is there a unit test that acceptorCounts is incremented and decremented correctly?

Reply:
> These aren't called when returning on line 303. Should they be in a finally?

By design. This should only dispose when the while loop is broken (like it was before). AFAIK every other piece of code has a try catch and a continue. This code path schedules itself for execution to the threadpool (which will restart the loop).

> Is there a unit test that acceptorCounts is incremented and decremented correctly?

Doubtful, this is difficult to test. None of the primitives used by this code are even mockable.